### PR TITLE
Bit 102 fix unbounded nonce dictionary

### PR DESCRIPTION
**Pull request by isabella618033**
#7 Saving and loading nonce dictionary upon restart of server 
#5 Nonce should always be larger then previous stored nonce and smaller then first nonce + 10 years of time
#3 Removing insecure prints from keyfile error logging

### CODE REVIEW

File: bittensor/_axon/__init__.py
Comment (camfairchild) [inline bittensor/_axon/__init__.py]:
```suggestion
                raise Exception(f"Nonce {nonce} is smaller or equal to previous nonce {previous_nonce} or larger than {first_nonce + 3.154e+17}.")
```

**Review by camfairchild — APPROVED** [review]
LGTM

File: bittensor/_axon/__init__.py
Comment (isabella618033) [reply bittensor/_axon/__init__.py]: oh nooooooo

File: bittensor/_axon/__init__.py
Comment (eduardogr) [inline bittensor/_axon/__init__.py]: As far as i know `is` is more encouraged that `==` or `!=` for this cases. 
Take a look at the references, I think it's a good scenario to compare using `is` instead of `==` 

I leave you here some references to that:
- https://stackoverflow.com/questions/14247373/python-none-comparison-should-i-use-is-or
- https://codeql.github.com/codeql-query-help/python/py-test-equals-none/

File: bittensor/_axon/__init__.py
Comment (eduardogr) [inline bittensor/_axon/__init__.py]: same here

File: bittensor/_axon/__init__.py
Comment (eduardogr) [inline bittensor/_axon/__init__.py:34]: I know that you did not change this in this PR, but as I mentioned in other PR,
What do you think about changing this absolute import to a relative one since we are inside the `bittensor` module

File: tests/unit_tests/bittensor_tests/test_axon.py
Comment (eduardogr) [inline tests/unit_tests/bittensor_tests/test_axon.py]: I know that we are not doing this in so many different tests but...I am going to share my point of view with this kind of practices in terms of unit testing.

In order to test this  piece of code it is pointed out as good practices to avoid to manage exceptions and also that we should not allow to have side effects generally, like the creation of the file by the code being tested.
In order to do it so, we should mock that piece of code so the file is not created, so we do not have to remove it. Tests should be run in a way that no side effects are produced.

What do you think about this?
If you agree and like this, feel free to reach me out to address this together in case you do not know how to proceed there.

**Comment by eduardogr** [comment]
left some comments, thanks for this 👍🏽

File: bittensor/_axon/__init__.py
Comment (isabella618033) [reply bittensor/_axon/__init__.py:34]: Hey, @eduardogr cause the same practise have been widely used across the project, so i think it might be better if we can change them in a separate PR.

Cause axon is a rather big submodule and it is using a lots of other bittensor submodules like prometheus /  wallet / proto / threadpool / config / grpc.

File: tests/unit_tests/bittensor_tests/test_axon.py
Comment (isabella618033) [reply tests/unit_tests/bittensor_tests/test_axon.py]: The file was created from the [test_saving](https://github.com/opentensor/bittensor/pull/1253/files/68326b438dcecf317a9f1d1ec916e7245e6a3c61#diff-9bb2374a00ee2f68a526e31be4b02d7c663f17bad6e67b2693194fa7f4852a08R1287) and [test_loading](https://github.com/opentensor/bittensor/pull/1253/files/68326b438dcecf317a9f1d1ec916e7245e6a3c61#diff-9bb2374a00ee2f68a526e31be4b02d7c663f17bad6e67b2693194fa7f4852a08R1296) function that is targeting at the [saving](https://github.com/opentensor/bittensor/pull/1253/files#diff-369f767fecee3a55495f39d9f75d496a6a61cac45c0ea2491760d190d38695d3R393) and [loading](https://github.com/opentensor/bittensor/pull/1253/files/68326b438dcecf317a9f1d1ec916e7245e6a3c61#diff-369f767fecee3a55495f39d9f75d496a6a61cac45c0ea2491760d190d38695d3R386) function here which are some basic file reading/writing function. So I am not sure how can we test it with mock, or is it actually not necessary to get these 2 function tested?

File: bittensor/_axon/__init__.py
Comment (eduardogr) [reply bittensor/_axon/__init__.py:34]: ok, i see, fair enough

File: tests/unit_tests/bittensor_tests/test_axon.py
Comment (eduardogr) [reply tests/unit_tests/bittensor_tests/test_axon.py]: yeah I know it's beign created. In terms of logic this is Ok. For the sake of simplicity and merge this, if you want, just proceed as it is, it's fine.

From the testing point of view, when testing that code that creates the file, we should mock that so no file is created. And we just intereact with the Mock. If you want just merge as it is and let's create a tech-deb task for that

**Review by eduardogr — APPROVED** [review]

File: tests/unit_tests/bittensor_tests/test_axon.py
Comment (isabella618033) [reply tests/unit_tests/bittensor_tests/test_axon.py]: Thanks edu, i understand, let me change the test to mock before merging it:)

**Pull request merged**
